NFC user dict list, delete, and de-duplication.

djsime1 opened the pull request:
# What's new

- Mifare Classic user keys can be listed and deleted.
- Duplicate keys (from user DB) can't be added to user database.
- A few minor changes to icon names.

# Verification 

- Open Mifare Classic Keys menu in NFC
- If user dict is not empty, list button (right) will be active.
- Add a few keys, then select one in the list to delete.
- Verify the correct key was deleted.
- Try to add a key already present in user database, and get informed it already exists.

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

djsime1 committed:
Add MFC user keys list

djsime1 committed:
Leakey submenu fix

djsime1 committed:
Set next target for Save/Delete success scenes

djsime1 committed:
Merge branch 'dev' into user-dict-list

djsime1 committed:
Delete individual user keys

djsime1 committed:
Update count of total keys

djsime1 committed:
Merge branch 'dev' into user-dict-list

djsime1 committed:
Fix memory leak

djsime1 commented on applications/nfc/scenes/nfc_scene_mf_classic_keys_list.c:
```diff
@@ -59,6 +69,12 @@ bool nfc_scene_mf_classic_keys_list_on_event(void* context, SceneManagerEvent ev
     Nfc* nfc = context;
-    // TODO: Find some way to clear all those strings from submenu.
+
+    MfClassicUserKeys_it_t it;
```
Memory leak was mostly fixed with latest commit, however not completely; I suspect it has to do with this array iterator.

djsime1 committed:
Check for duplicate keys

djsime1 committed:
Remove a submodule that I never added?

djsime1 committed:
Format

djsime1 committed:
Merge branch 'dev' into user-dict-list

gornekich commented:
Hello @djsime1. Thank you for the PR!

Dealing with duplicated keys is a good idea. Also our UX/UI team liked your idea with User dictionary keys listing and deleting. However, we want to add this a little bit differently. Our designer is working on UI for this feature right now and soon it will be in public Miro https://miro.com/app/board/uXjVOlznagk=/?share_link_id=903187096944

I will let you know when the design is ready. Please tell me if you want to rework this PR once design is finalized.

gornekich commented:
Also some changes came to dev recently. Please, sync  the PR with latest dev

djsime1 committed:
Merge branch 'dev' into user-dict-list

djsime1 commented:
Got dev merged, I have limited availability the next few weeks but I'll keep an eye out for that UI/UX being published.

djsime1 committed:
Merge branch 'dev' into user-dict-list

djsime1 committed:
Swap and position icons

djsime1 commented:
Jumped up to latest dev and fixed issues caused by icon asset changes.

gornekich commented:
Hello @djsime1 
We updated user dictionary flow here https://miro.com/app/board/uXjVOlznagk=/
Please, tell me if you will implement it in this PR

djsime1 commented:
> Hello @djsime1 We updated user dictionary flow here https://miro.com/app/board/uXjVOlznagk=/ Please, tell me if you will implement it in this PR

Yes, I can implement that flow. I'll go with the top "Key exists" screen since I don't know how to draw a canvas atop the byteinput. Even if I did, switching to a popup scene is still simpler.

![image](https://user-images.githubusercontent.com/8518150/186947871-15633ad1-ecd1-4225-962c-498e46482b94.png)

djsime1 commented:
Hopefully I'll have this ready by tomorrow night.

djsime1 commented:
@gornekich I'm almost ready to push my changes, however Miro won't let me download this icon in its raw format. Could you please reply with the icon attached? Thanks.
![image](https://user-images.githubusercontent.com/8518150/187274602-0c50f60a-a84a-4a2c-9542-bd007918f49b.png)

djsime1 committed:
Revamp according to design doc

djsime1 committed:
Merge branch 'dev' into user-dict-list

gornekich commented:
@djsime1 I added this asset myself, it's already in dev. 
Is your PR ready for review?

djsime1 commented:
> @djsime1 I added this asset myself, it's already in dev. Is your PR ready for review?

The [keychain icon currently present in dev](https://github.com/flipperdevices/flipperzero-firmware/blob/dev/assets/icons/NFC/Keychain.png) is not the same as the Miro keychain icon. The dev one is significantly larger, resulting in the scenes title being truncated from "Mifare Classic Keys" to "MF Classic Keys." ~~Additionally, I think there's one tiny bug about pressing LEFT in the delete confirmation screen.~~

Besides that, it is ready for review. My next commit will ~~fix that LEFT bug,~~ as well as the smaller keychain icon if I receive it before the commit is pushed.

_EDIT: Forgot I already fixed the bug with previous commit, so the icon is the only thing that's left._ **Ready for review!**

djsime1 committed:
Merge branch 'dev' into user-dict-list

Karator commented:
Hi bro!
We attach that new picture here.
![Small_keychain_36x39](https://user-images.githubusercontent.com/12154975/187484355-e916e889-7951-4968-8395-700e6b2d1107.png)

It was pinned to Miiro board but sometimes it's tricky to download it ((

djsime1 committed:
Rename icons to include size and replace keychain icon with smaller variant

djsime1 committed:
Fix typos

djsime1 commented:
Wrong again I was! The confirmation cancel bug was an odd quirk with the order in which scene loading is handled. Long story short, MFC dict was allocated before the if statement, and free'd at the end of the event callback function. Little did I know the current scenes exit and next scenes enter functions were called in between. Sure enough, the next scene's MFC dict allocate statement was called before our free statement. One last commit and this PR will be finished.

djsime1 committed:
Final fixes

djsime1 commented:
Alright, everything's ready to go!

gornekich commented on applications/nfc/scenes/nfc_scene_mf_classic_keys_list.c:
```diff
@@ -0,0 +1,66 @@
+    MfClassicUserKeys_it_t it;
+    for(MfClassicUserKeys_it(it, mfc_key_strs); !MfClassicUserKeys_end_p(it);
+        MfClassicUserKeys_next(it)) {
+        free(MfClassicUserKeys_ref(it));
```
Memory leak here. Must be free(*MfClassicUserKeys_ref(it));

gornekich commented on applications/nfc/scenes/nfc_scene_mf_classic_keys_list.c:
```diff
@@ -0,0 +1,66 @@
+
+ARRAY_DEF(MfClassicUserKeys, char*, M_PTR_OPLIST);
+
+static MfClassicUserKeys_t mfc_key_strs;
```
Really don't like to see static variable here. I think that you can add this array instance in Nfc structure in nfc_i.h, so that it will be dynamically  allocated along with nfc control structure

gornekich reviewed (changes requested):
@djsime1 you made a great work, thank you!
Could you please add a few changed and we will merge you PR

djsime1 commented:
Sure, I'll get to work on that when I can.

djsime1 committed:
Merge branch 'dev' into user-dict-list

djsime1 committed:
Fufill requested changes

djsime1 committed:
Cleanup comments

djsime1 committed:
Merge dev after SD app loading

djsime1 committed:
Merge branch 'dev' into user-dict-list

djsime1 committed:
Fixing icon names

djsime1 commented:
Could someone please review/merge this commit? I've not gotten any feedback in the past week. Thanks!

skotopes commented:
I’ll take a look today. We’ve been waiting for elf support to be merged first.

gornekich commented on applications/main/nfc/scenes/nfc_scene_delete.c:
```diff
@@ -59,6 +59,8 @@ bool nfc_scene_delete_on_event(void* context, SceneManagerEvent event) {
         } else if(event.event == GuiButtonTypeRight) {
             if(nfc_device_delete(nfc->dev, true)) {
+                scene_manager_set_scene_state(
+                    nfc->scene_manager, NfcSceneDeleteSuccess, NfcSceneStart);
```
This breaks current logic. We want to switch to NfcSceneFileSelect after delete success.

gornekich commented on applications/main/nfc/scenes/nfc_scene_delete_success.c:
```diff
@@ -26,7 +26,8 @@ bool nfc_scene_delete_success_on_event(void* context, SceneManagerEvent event) {
             consumed = scene_manager_search_and_switch_to_previous_scene(
-                nfc->scene_manager, NfcSceneFileSelect);
+                nfc->scene_manager,
+                scene_manager_get_scene_state(nfc->scene_manager, NfcSceneDeleteSuccess));
```
I suggest different approach here. You search for previous scene in scene manager stack and choose where to go next. Something like that:
if(scene_manager_has_previous_scene(nfc->scene_manager, NfcSceneMfClassicKeysAdd) {
    scene_manager_search_and_switch_to_previous_scene(nfc->scene_manager, NfcSceneMfClassicKeys)
} else {
    scene_manager_search_and_switch_to_previous_scene(
                nfc->scene_manager, NfcSceneFileSelect);
}

gornekich commented on applications/main/nfc/scenes/nfc_scene_dict_not_found.c:
```diff
@@ -31,13 +31,9 @@ bool nfc_scene_dict_not_found_on_event(void* context, SceneManagerEvent event) {
-            }
+            consumed = scene_manager_search_and_switch_to_previous_scene(
+                nfc->scene_manager,
+                scene_manager_get_scene_state(nfc->scene_manager, NfcSceneDictNotFound));
```
Same here. I think it's not good when one scene state is changed from another scene. It's not how scene states are designed to be used

gornekich commented on applications/main/nfc/scenes/nfc_scene_mf_classic_keys.c:
```diff
@@ -47,6 +59,14 @@ bool nfc_scene_mf_classic_keys_on_event(void* context, SceneManagerEvent event)
+            consumed = true;
+        } else if(
+            event.event == GuiButtonTypeRight &&
+            (scene_manager_get_scene_state(nfc->scene_manager, NfcSceneMfClassicKeys) > 0)) {
```
There is no need in saving number of user keys in scene state. Once you get user_dict_keys_total in _on_enter() handler, you add button element to widget. If you user_dict_keys_total = 0 and you don't add button element, you will never get GuiButtonRight event

djsime1 replied:
I didn't know the event only fired when the button was present.

gornekich commented on applications/main/nfc/scenes/nfc_scene_mf_classic_keys.c:
```diff
@@ -23,18 +23,30 @@ void nfc_scene_mf_classic_keys_on_enter(void* context) {
     if(dict) {
         user_dict_keys_total = mf_classic_dict_get_total_keys(dict);
         mf_classic_dict_free(dict);
+        scene_manager_set_scene_state(
```
Not necessary

gornekich reviewed (changes requested):
Please, also revert changes in readme files. Current changes seem strange.
Also, since you changed icon names, our build system increased API version. We don't want to increment major API version for such small changes for now, could you please change API version back to 1.0 by hand. This can be done in api_symbols.csv file.

You can also give maintainers access to push to you branch so we can help to resolve some issues, if you don't mind

djsime1 commented:
Readme changes must have been added by the merge, since I hadn't touched the file at all. I was confused at how the API symbols system worked so I just let it do its thing, but I'll change it back to 1.0. Maintainer push access has been enabled since I originally opened the PR. Lastly, my though process for popup scenes was that you'd be able to dictate where they return too if needed instead of always going to file select. I guess I can undo my changes to the deleted popup since a MFC-Key specific delete scene was added.

djsime1 committed:
Revert merge mistakes and API version

djsime1 committed:
Scene switching adjustments

djsime1 commented:
Done.

gornekich reviewed (dismissed):
Thanks!

djsime1 committed:
Merge branch 'dev' into user-dict-list

skotopes committed:
Merge branch 'dev' into user-dict-list

skotopes committed:
F7: add/change/remove some nfc icons in api_symbols.csv
